### PR TITLE
Improve gulp file

### PR DESCRIPTION
**Pull request by faf**

### CODE REVIEW

File: src/gulpfile.js [hunk 1]
```diff
@@ -186,47 +186,74 @@ gulp.task('chat-styles', ['chat-styles-handlebars', 'chat-styles-js'], function(
...
-            handlebars: handlebarsEngine
-        }))
-        .pipe(wrapHandlebarsTemplate())
-        .pipe(concat('templates.js'))
-        .pipe(uglify({preserveComments: 'some'}))
-        .pipe(header(config.compiledTemplatesHeader))
-        .pipe(gulp.dest(stylePath + '/templates_compiled/client_side'));
+    var promises = [];
+    fs.readdir(config.chatStylesPath, function(err, list){
+        if(err) return done(err);
```
Comment (JustBlackBird) [inline src/gulpfile.js]: `done` is not a function

File: src/gulpfile.js [hunk 2]
```diff
@@ -186,47 +186,74 @@ gulp.task('chat-styles', ['chat-styles-handlebars', 'chat-styles-js'], function(
...
+        list.filter(function(path){
+            return fs.lstatSync(config.chatStylesPath + "/" + path).isDirectory();
+        }).map(function(dir){
+            promises.push(new Promise(function(resolve, reject){
```
Comment (JustBlackBird) [inline src/gulpfile.js]: `Array.prototype.push` is not needed. Just return a promise from (that's `Array.prototype.map` is designed for).

File: src/gulpfile.js [hunk 3]
```diff
@@ -186,47 +186,74 @@ gulp.task('chat-styles', ['chat-styles-handlebars', 'chat-styles-js'], function(
...
-        .pipe(handlebars({
-            // Use specific version of Handlebars.js
-            handlebars: handlebarsEngine
-        }))
-        .pipe(wrapHandlebarsTemplate())
-        .pipe(concat('templates.js'))
-        .pipe(uglify({preserveComments: 'some'}))
-        .pipe(header(config.compiledTemplatesHeader))
-        .pipe(gulp.dest(stylePath + '/templates_compiled/client_side'));
+    var promises = [];
```
Comment (JustBlackBird) [inline src/gulpfile.js]: All promises will be in this array but what's next? Remeber: the task is async, so you have to say Gulp when it's really ends (or fails).

File: src/gulpfile.js [hunk 4]
```diff
@@ -186,47 +186,74 @@ gulp.task('chat-styles', ['chat-styles-handlebars', 'chat-styles-js'], function(
...
+    fs.readdir(config.chatStylesPath, function(err, list){
+        if(err) return done(err);
+        list.filter(function(path){
+            return fs.lstatSync(config.chatStylesPath + "/" + path).isDirectory();
```
Comment (JustBlackBird) [inline src/gulpfile.js]: Sync methods are evil 😈

File: src/gulpfile.js [hunk 5]
```diff
@@ -186,47 +186,74 @@ gulp.task('chat-styles', ['chat-styles-handlebars', 'chat-styles-js'], function(
...
+                  .pipe(handlebars({
+                      // Use specific version of Handlebars.js
+                      handlebars: handlebarsEngine
+                  }))
+                  .pipe(wrapHandlebarsTemplate())
+                  .pipe(concat('templates.js'))
+                  .pipe(uglify({preserveComments: 'some'}))
+                  .pipe(header(config.compiledTemplatesHeader))
+                  .pipe(gulp.dest(config.chatStylesPath + '/' + dir + '/templates_compiled/client_side'));
+                pipeline.on('end', resolve);
```
Comment (JustBlackBird) [inline src/gulpfile.js]: As far as I know, `EventEmitter.prototype.on` methods could be chained, so there is no need for extra variable `pipeline`.

**Comment by faf** [comment]
@snewell92 Go on. :)

**Comment by faf** [comment]
Actually at the moment I'm very intrigued by the fact that the aforementioned code actually works.

**Comment by snewell92** [comment]
[Here is the PR](https://github.com/Mibew/mibew/pull/196) that addresses the review comments, and removes the unused promises. If those were used for a reason, I'd like to know how and why. The other gulp tasks don't use them.

The other task functions also return a pipe(?) so maybe it should do that? Please add comments as you see fit to the PR.

**Pull request merged**